src/App.spec.ts

```diff
@@ -342,18 +344,147 @@ describe('App', () => {
 
         // Assert
         assert.instanceOf(fakeErrorHandler.firstCall.args[0], Error);
       });
+
+      it('correctly waits for async listeners', async () => {
+        let changed = false;
+
+        app.use(async ({ next }) => {
+          await delay(100);
+          changed = true;
+
+          await next();
+        });
+
+        await fakeReceiver.sendEvent(dummyReceiverEvent);
+        assert.isTrue(changed);
+        assert(fakeErrorHandler.notCalled);
+      });
+
+      it('throws errors which can be caught by upstream async listeners', async () => {
+        const thrownError = new Error('Error handling the message :(');
+        let caughtError;
+
+        app.use(async ({ next }) => {
+          try {
+            await next();
+          } catch (err) {
+            caughtError = err;
+          }
+        });
+
+        app.use(async () => {
+          throw thrownError;
+        });
+
+        app.error(fakeErrorHandler);
+
+        await fakeReceiver.sendEvent(dummyReceiverEvent);
+
+        assert.equal(caughtError, thrownError);
+        assert(fakeErrorHandler.notCalled);
+      });
+
+      it('calls async middleware in declared order', async () => {
+        const message = ':wave:';
+        let middlewareCount = 0;
+
+        const assertOrderMiddleware = (orderDown: number, orderUp: number) => async ({ next }: any) => {
+          await delay(100);
+          middlewareCount += 1;
+          assert.equal(middlewareCount, orderDown);
+          await next();
+          middlewareCount += 1;
+          assert.equal(middlewareCount, orderUp);
+        };
+
+        app.use(assertOrderMiddleware(1, 8));
+        app.message(message, assertOrderMiddleware(3, 6), assertOrderMiddleware(4, 5));
+        app.use(assertOrderMiddleware(2, 7));
+        app.error(fakeErrorHandler);
+
+        await fakeReceiver.sendEvent({
+          ...dummyReceiverEvent,
+          body: {
+            type: 'event_callback',
+            event: {
+              type: 'message',
+              text: message,
+            },
+          },
+        });
+
+        assert.equal(middlewareCount, 8);
+        assert(fakeErrorHandler.notCalled);
+      });
+
+      it('should, on error, ack an error, then global error handler', async () => {
+        const error = new Error('Everything is broke, you probably should restart, if not then good luck');
+        let callCount = 0;
+
+        const assertOrder = (order: number) => {
+          callCount += 1;
+          assert.equal(callCount, order);
+        };
+
+        app.use(() => {
+          assertOrder(1);
+          throw error;
+        });
+
+        const event = {
+          ...dummyReceiverEvent,
+          ack: (actualError: Error): Promise<void> => {
+            if (actualError instanceof Error) {
+              assert.instanceOf(actualError, UnknownError);
+              assert.equal(actualError.message, error.message);
+              assertOrder(2);
```

Comment:
note: this will be updated once we make the change where `ack()` does not receive errors.

Reply:
👍 Rewrote this to simply assert that the global error handler is called.